lib/libbackscrub.cc

```diff
@@ -0,0 +1,270 @@
+// label number of "person" for DeepLab v3+ model
+static const size_t cnum = labels.size();
+static const size_t pers = std::distance(labels.begin(), std::find(labels.begin(),labels.end(),"person"));
+
+int init_tensorflow(calcinfo_t &info) {
+	// Deallocate if not done already
+	if (info.backscrub_ctx)
+		drop_tensorflow(info);
+	// Allocate context
+	info.backscrub_ctx = new backscrub_ctx_t;
+	// Take a reference so we can write tidy code with ctx.<x>
+	backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
+	// Load model
+	ctx.model = tflite::FlatBufferModel::BuildFromFile(info.modelname);
+	TFLITE_MINIMAL_CHECK(ctx.model != nullptr);
+	// Determine model type and normalization values
+	ctx.modeltype = get_modeltype(info.modelname);
+	ctx.norm = get_normalization(ctx.modeltype);
+	if (modeltype_t::Unknown == ctx.modeltype) {
+		_dbg(info, "Unknown model type '%s'.\n", info.modelname);
+		return 0;
+	}
+	// Build the interpreter
+	tflite::ops::builtin::BuiltinOpResolver resolver;
+	// custom op for Google Meet network
+	resolver.AddCustom("Convolution2DTransposeBias", mediapipe::tflite_operations::RegisterConvolution2DTransposeBias());
+	tflite::InterpreterBuilder builder(*ctx.model, resolver);
+	builder(&ctx.interpreter);
+	TFLITE_MINIMAL_CHECK(ctx.interpreter != nullptr);
+
+	// Allocate tensor buffers.
+	TFLITE_MINIMAL_CHECK(ctx.interpreter->AllocateTensors() == kTfLiteOk);
+
+	// set interpreter params
+	ctx.interpreter->SetNumThreads(info.threads);
+	ctx.interpreter->SetAllowFp16PrecisionForFp32(true);
+
+	// get input and output tensor as cv::Mat
+	info.input = getTensorMat(info, ctx.interpreter->inputs ()[0]);
+	info.output = getTensorMat(info, ctx.interpreter->outputs()[0]);
+	info.ratio = (float)info.input.cols/(float) info.input.rows;
+
+	// initialize mask and square ROI in center
+	info.roidim = cv::Rect((info.width-info.height/info.ratio)/2,0,info.height/info.ratio,info.height);
+	info.mask = cv::Mat::ones(info.height,info.width,CV_8UC1)*255;
+	info.mroi = info.mask(info.roidim);
+
+	// mask blurring size
+	info.blur = cv::Size(5,5);
+
+	// create Mat for small mask
+	info.ofinal = cv::Mat(info.output.rows,info.output.cols,CV_8UC1);
+	return 1;
+}
+
+void drop_tensorflow(calcinfo_t &info) {
+	if (info.debug)
+		_dbg(info, "dropping tensorflow context\n");
+	// clear all mask data
+	info.ofinal.deallocate();
+	info.mask.deallocate();
+	info.input.deallocate();
+	info.output.deallocate();
+	// clear internal context if present
+	if (info.backscrub_ctx) {
+		backscrub_ctx_t &ctx = *((backscrub_ctx_t *)info.backscrub_ctx);
+		// drop interpreter
+		ctx.interpreter.reset();
+		// drop model
+		ctx.model.reset();
+		// drop context
+		delete &ctx;
+		info.backscrub_ctx = nullptr;
```

Comment:
Race condition with potential UAF. Always first clear externally visible references, then release associated memory.

Reply:
Agreed, and another case of removing the implementation detail from callers first, then fixing what remains.